### PR TITLE
feat(llmhttp): route HTTP/1.1 upstreams through fasthttp

### PR DESCRIPTION
**Pull request by invakid404**
## Summary

- Adds a fasthttp backend behind the existing `llmhttp.Client` API — callers (orchestrators, adapter) stay unchanged.
- Per-origin ALPN probe decides net/http (h2) vs fasthttp (h1); cached lock-free via atomic.Pointer with singleflight-gated first-hits.
- `BAML_REST_HTTP_CLIENT=auto|fasthttp|nethttp` overrides the probe. `fasthttp` skips probing entirely for internal-service deployments that know their upstream is h1.
- `http://` short-circuits to fasthttp immediately (no probe). `https://` behind a configured proxy pins to net/http (fasthttp has no `ProxyFromEnvironment`).

## Design notes

**Streaming path uses a per-request HostClient with a custom Dial** that captures the net.Conn into a slot. On ctx cancel the slot closes the socket directly; fasthttp's normal cleanup then runs via `CloseBodyStream`. We do *not* use `CloseWithError` on the streaming path — it calls `releaseRequestStream` concurrently with an in-flight `requestStream.Read`, which the race detector correctly flags as a pool-corruption hazard.

**Non-streaming path uses the pooled HostClient** (so we keep conn reuse at the hot HTTP/1.1-at-1000s-RPS case). `DoDeadline` installs a per-conn `SetReadDeadline` equal to `ctx.Deadline()`, so Read usually unblocks naturally before we observe `ctx.Done()`. The ctx-cancel path waits up to 500 ms for the Read goroutine to exit naturally before falling back to `CloseWithError`, eliminating the race in the common case (cancel-via-timeout).

**SetNoDefaultContentType(true)** is set after each streaming `Do`: fasthttp's `ContentType()` synthesises a `text/plain; charset=utf-8` default when the response has none, which would mask the distinct "missing Content-Type" error path. Must be called after `Do` because `resp.Reset()` inside `Do` clears the flag.

**Response header snapshot** is taken before handing the body stream to `sseclient`: `requestStream.Read` mutates the header (trailer parsing) concurrently with any later `VisitAll`.

**`TestError_MidStreamDisconnect`** is pinned to net/http mode (documented in the test). fasthttp's streaming chunked reader treats a between-chunks EOF as a clean stream end — a genuine behaviour difference from net/http's `ErrUnexpectedEOF`. Not a regression, not fixable without patching fasthttp.

**Startup warmup is intentionally out of scope.** `introspected.ClientProvider` only exposes provider names; base URLs are runtime-only (from `__baml_options__.ClientRegistry` + BAML-SDK-internal env substitution). For the primary use case (known-h1 internal service), `BAML_REST_HTTP_CLIENT=fasthttp` skips the probe entirely — no cold-start cost. For `auto` mode with HTTPS, the first request per host pays one TLS handshake.

## Test plan

- [x] `go test -race ./bamlutils/...` (clean)
- [x] `go test -race -count=30 ./bamlutils/llmhttp/...` (clean)
- [x] `go test -race -count=30 ./bamlutils/buildrequest/...` (clean)
- [ ] CI `go test -race -count=100 ./...` — will confirm no residual races
- [ ] Manual: drive a request through `BAML_REST_HTTP_CLIENT=fasthttp` and verify the hot path (internal h1 service)
- [ ] Manual: verify `auto` mode against a real h2 provider (OpenAI) pins to net/http after first response

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added an alternative fast HTTP backend with per-origin routing and a configurable client mode, improving streaming (SSE) and non-streaming request performance, cancellation, and response-size enforcement.

* **Bug Fixes**
  * Improved origin normalization, Host header handling, proxy behavior, and robustness around truncated SSE streams and error propagation.

* **Tests & CI**
  * Expanded tests covering client modes, SSE, edge cases and concurrency; CI matrix now exercises multiple HTTP client configurations and extended unit-test runtime.

* **Chores**
  * Updated embedded resources and project dependency declarations.

### CODE REVIEW

**Comment by invakid404** [comment]
@coderabbitai full review

File: bamlutils/llmhttp/fast_backend.go [hunk 1]
```diff
@@ -0,0 +1,781 @@
...
+
+	select {
+	case <-ctx.Done():
+		// Hand off cleanup so the pool isn't starved by the orphaned Do.
+		go func() {
+			<-doneCh
+			fasthttp.ReleaseRequest(fReq)
+			fasthttp.ReleaseResponse(fResp)
+		}()
+		return nil, ctx.Err()
```
[the remark this answers is not in the document]
Comment (invakid404) [reply bamlutils/llmhttp/fast_backend.go:56]: Not applicable. `fasthttp.ReleaseResponse` calls `Response.Reset()` which calls `ResetBody()` → `closeBodyStream()`. The body stream is properly closed before release. Verified against fasthttp v1.70.0 source (`client.go:1244` → `http.go:1290` → `http.go:777`).

File: bamlutils/llmhttp/fast_backend.go [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+	}
+
+	bodyStream := fResp.BodyStream()
+	if bodyStream == nil {
+		// Defensive: without StreamResponseBody on the HostClient the body
+		// would be pre-buffered and BodyStream would be nil. That would be
+		// a configuration bug in this package.
+		releaseFastExchange(fReq, fResp)
+		return nil, fmt.Errorf("llmhttp: fasthttp response has no body stream (is StreamResponseBody enabled?)")
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply bamlutils/llmhttp/fast_backend.go:229]: The duplicate comment (error branches at line 191-195) describes the same issue. Addressing in the next commit.

File: bamlutils/llmhttp/llmhttp.go [hunk 3]
```diff
@@ -151,8 +181,26 @@ func (c *Client) ExecuteStream(ctx context.Context, req *Request) (*StreamRespon
...
+	rewritten := resolveRequestURL(req)
 
-	httpReq, err := buildHTTPRequest(ctx, req)
+	// Dispatch to fasthttp when the per-origin cache says the host speaks
+	// HTTP/1.1. Cache misses probe (or short-circuit on http://) before
+	// returning; cache hits are a single atomic load.
+	if c.cache != nil {
+		if origin, err := parseOrigin(rewritten); err == nil {
+			if entry := c.cache.resolve(origin); entry != nil && entry.decision == decisionFast && entry.host != nil {
+				return c.executeStreamFast(ctx, req, rewritten, entry.host)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply bamlutils/llmhttp/llmhttp.go:229]: Fixed in `52307dd31`. `resolve()` now accepts a `ctx` parameter. Uses `singleflight.DoChan` with `select { <-ch; <-ctx.Done() }` so canceled requests return immediately while the probe continues in the background for other waiters. Probe TLS dial uses `tls.Dialer.DialContext` with bounded timeout.

File: bamlutils/llmhttp/llmhttp.go [hunk 4]
```diff
@@ -87,20 +87,50 @@ const MaxResponseBodyBytes = 16 << 20 // 16 MiB
...
+// well as the net/http backend.
 func NewClient(httpClient *http.Client) *Client {
 	if httpClient == nil {
 		httpClient = http.DefaultClient
 	}
-	return &Client{httpClient: httpClient}
+	return &Client{
+		httpClient: httpClient,
+		cache:      newProtocolCache(loadClientMode(), tlsConfigFromHTTPClient(httpClient)),
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply bamlutils/llmhttp/llmhttp.go:135]: By design — fasthttp IS the alternative to `net/http`, not a wrapper around it. Documented in `52307dd31`: `NewClient` now explicitly states that when fasthttp is selected, only TLS config and proxy detection carry over from the injected `http.Client`. `Timeout`, `CheckRedirect`, `Jar`, and custom Transport wrappers are not honored. Callers needing those should force `nethttp` mode via `BAML_REST_HTTP_CLIENT=nethttp`.

File: bamlutils/llmhttp/fast_backend_test.go [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+	}
+	h := <-seen
+	if h.Get("Authorization") != "Bearer test" {
+		t.Errorf("Authorization not forwarded: %q", h.Get("Authorization"))
+	}
+	// Go's http server normalises keys to canonical form on read; the
+	// point here is just that the value round-tripped.
+	if h.Get("X-Custom") != "v1" {
+		t.Errorf("x-custom not forwarded: %q", h.Get("X-Custom"))
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply bamlutils/llmhttp/fast_backend_test.go:226]: Fixed in `52307dd31`. Rewrote the test to use a raw `net.Listener` that reads HTTP bytes off the wire and asserts exact header-name casing. This also caught a latent bug: `HostClient.DisableHeaderNamesNormalizing` only affects response parsing, not request writes. Added `fReq.Header.DisableNormalizing()` in `buildFastRequest` to preserve caller header casing on the wire.

File: bamlutils/llmhttp/fast_backend.go [hunk 6]
```diff
@@ -0,0 +1,795 @@
...
+					host, _, splitErr := net.SplitHostPort(addr)
+					if splitErr == nil {
+						tlsCfg.ServerName = host
+					}
+				}
+				tlsConn := tls.Client(rawConn, tlsCfg)
+				if err := tlsConn.HandshakeContext(ctx); err != nil {
+					_ = rawConn.Close()
+					return nil, err
+				}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply bamlutils/llmhttp/fast_backend.go:427]: Not applicable. `tmpl.TLSConfig` is guaranteed non-nil at this call site — `newProtocolCache` (`cache.go:108`) forces a non-nil TLS config at construction, and `buildEntry` (`cache.go:313`) copies it into every `HostClient` template. There is no production path where `tmpl.TLSConfig` is nil when this code runs.

File: bamlutils/llmhttp/fast_backend.go [hunk 7]
```diff
@@ -0,0 +1,795 @@
...
+		return nil, ctx.Err()
+	case doErr = <-doneCh:
+	}
+	// Pre-Do watcher has served its purpose — the post-Do reader wires its
+	// own ctx watcher on the same slot.
+	cancelPreDo()
+	if doErr != nil {
+		fasthttp.ReleaseRequest(fReq)
+		fasthttp.ReleaseResponse(fResp)
+		return nil, fmt.Errorf("llmhttp: request failed: %w", doErr)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply bamlutils/llmhttp/fast_backend.go:179]: Duplicate — already addressed in `52307dd31`. The `teardownOnError` closure now runs `releaseFastExchange` + `hc.CloseIdleConnections()` + `slot.shutdown()` for all early stream error branches.

File: .github/workflows/integration-tests.yml [hunk 8]
```diff
@@ -25,9 +25,13 @@ jobs:
...
           # Read the versions file and build matrix combining pinned + latest
-          # Cross with unary-server on/off to test both configurations
+          # Cross with unary-server on/off to test both configurations.
+          # http-client is crossed only with the BuildRequest path — the legacy
+          # CallStream path goes through the BAML runtime's own HTTP client,
+          # so BAML_REST_HTTP_CLIENT has no effect there and duplicating those
+          # runs would just burn CI minutes.
           VERSIONS=$(jq -c '[.pinned[], .latest] | unique' integration/baml_versions.json)
-          echo "matrix={\"baml-version\":$VERSIONS,\"unary-server\":[false,true],\"use-build-request\":[false,true]}" >> $GITHUB_OUTPUT
+          echo "matrix={\"baml-version\":$VERSIONS,\"unary-server\":[false,true],\"use-build-request\":[false,true],\"http-client\":[\"nethttp\",\"fasthttp\"],\"exclude\":[{\"use-build-request\":false,\"http-client\":\"fasthttp\"}]}" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (invakid404) [reply .github/workflows/integration-tests.yml]: Fixed in `61df6840c`. Added `"auto"` to both matrix blocks (versioned + BAML-source jobs). Excluded against `!use-build-request` same as `fasthttp`, since the legacy path does not exercise `llmhttp`.

File: bamlutils/llmhttp/cache.go [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+	if c.proxyFunc != nil {
+		req := &http.Request{
+			URL:    &url.URL{Scheme: origin.scheme, Host: origin.addr()},
+			Method: "GET",
+			Host:   origin.addr(),
+		}
+		if p, err := c.proxyFunc(req); err == nil && p != nil {
+			return decisionNet
+		}
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply bamlutils/llmhttp/cache.go:288]: Fixed in `61df6840c`. Proxy function errors now return `decisionNet` (safe fallback to net/http) instead of falling through to ALPN/fast routing.

File: bamlutils/llmhttp/cache.go [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+		} else {
+			port = "80"
+		}
+	}
+	return originURL{
+		key:      scheme + "://" + net.JoinHostPort(host, port),
+		scheme:   scheme,
+		hostname: host,
+		port:     port,
+	}, nil
```
[the remark this answers is not in the document]
Comment (invakid404) [reply bamlutils/llmhttp/cache.go:181]: Fixed in `61df6840c`. `parseOrigin` now lowercases the hostname alongside the scheme. Mixed-case origins share a single cache entry and HostClient. Added table-driven test entries for the regression.
